src/osd/scrubber/scrub_machine.h

```diff
@@ -596,45 +613,96 @@ struct WaitDigestUpdate : sc::state<WaitDigestUpdate, ActiveScrubbing>,
 
 // ----------------------------- the "replica active" states
 
-/**
- * ReservedReplica
+/*
+ *  The replica states:
+ *
+ *  ReplicaActive - starts after being peered as a replica. Ends on interval.
+ *   - maintain the "I am reserved by a primary" state;
+ *   - handles reservation requests
+ *
+ *     - ReplicaIdle - ready for a new scrub request
+ *          * initial state of ReplicaActive
  *
- * Parent state for replica states,  Controls lifecycle for
- * PgScrubber::m_reservations.
+ *     - ReplicaActiveOp - handling a single map request op
+ *          * ReplicaWaitUpdates
+ *  	    * ReplicaBuildingMap
  */
-struct ReservedReplica : sc::state<ReservedReplica, ScrubMachine, ReplicaIdle>,
+
+struct ReplicaIdle;
+
+struct ReplicaActive : sc::state<ReplicaActive, ScrubMachine, ReplicaIdle>,
 			 NamedSimply {
-  explicit ReservedReplica(my_context ctx);
-  ~ReservedReplica();
+  explicit ReplicaActive(my_context ctx);
+  ~ReplicaActive();
+
+  /// handle a reservation request from a primary
+  void on_reserve_req(const ReplicaReserveReq&);
+
+  /// handle a 'release' from a primary
+  void on_release(const ReplicaRelease&);
+
+  void check_for_updates(const StartReplica&);
 
-  using reactions = mpl::list<sc::transition<FullReset, NotActive>>;
+  using reactions = mpl::list<
+      // a reservation request from the primary
+      sc::in_state_reaction<
+	  ReplicaReserveReq,
+	  ReplicaActive,
+	  &ReplicaActive::on_reserve_req>,
+      // an explicit release request from the primary
+      sc::in_state_reaction<
+	  ReplicaRelease,
```

Comment:
This should re-enter ReplicaActive rather than use a separate in-state reaction.  We want ~ReplicaActive to be, truly, the only place responsible for releasing resources associated with ReplicaActive.

Reply:
@athanatos 

There are two issues with doing that (which is why I used an in-state reaction):

- conceptually - we've set out to decouple the handling of scrub requests from the reservations. And what you suggest means aborting an on-going scrub op on a 'release' message.  And:

- we do not currently have a way to signal a replica failure in handling the scrub request (apart from a 'was preempted' flag(*) ). I've already planned  to ask you about our handling of back-end errors (which I do not think are handled correctly now), but I did not plan to fix in this PR. So, back to our case - what should the behavior be if we get a release mid-op?

(one possible option to consider, if we are set on coupling the two: not aborting the current chunk, setting a flag, and handling the 'release' when the chunk ends.

On the plus side - the logic is clear. On the other hand - this complicates the state machine unnecessarily.)

--
(*) I also planned to discuss the scenarios in which we would encounter a preemption on a replica, but not on the primary.

Reply:
Ok, makes sense.

I think a release shouldn't actually affect an in progress op.  It shouldn't happen since the primary won't do it, but I don't think it's really worth adding special handling to notice it.  For backend errors, we already represent IO errors in the ScrubMap results.  I'd think anything more serious should probably just result in the death of the OSD, unless you had something specific in mind you'd like to handle differently?

I don't think it's possible to see a preemption on the replica that didn't happen first on the primary.  If the replica sees an IO during a scrub operation, I think the replica can safely drop it on the floor.  I suppose there could be an "I was preempted" message sent back to the primary, but in all cases the primary should already have been preempted.